src/Agent.Listener/Configuration/ConfigurationProvider.cs

```diff
@@ -0,0 +1,357 @@
+
+        void UpdateAgentSetting(AgentSettings settings);
+    }
+
+    public abstract class ConfigurationProvider : AgentService
+    {
+        public Type ExtensionType => typeof(IConfigurationProvider);
+        protected ITerminal _term;
+        protected IAgentServer _agentServer;
+
+        public override void Initialize(IHostContext hostContext)
+        {
+            base.Initialize(hostContext);
+            _term = hostContext.GetService<ITerminal>();
+        }
+
+        protected void InitializeServerConnection(IAgentServer agentServer)
+        {
+            _agentServer = agentServer;
+        }
+        
+        protected Task<TaskAgent> UpdateAgent(int poolId, TaskAgent agent)
+        {
+           return _agentServer.UpdateAgentAsync(poolId, agent);
+        }
+
+        protected Task<TaskAgent> AddAgent(int poolId, TaskAgent agent)
+        {
+            return _agentServer.AddAgentAsync(poolId, agent);
+        }
+
+        protected Task DeleteAgent(int poolId, int agentId)
+        {
+            return _agentServer.DeleteAgentAsync(poolId, agentId);
+        }
+
+        protected async Task TestConnectionAsync(string url, VssCredentials creds)
+        {
+            _term.WriteLine(StringUtil.Loc("ConnectingToServer"));
+            VssConnection connection = ApiUtil.CreateConnection(new Uri(url), creds);
+
+            _agentServer = HostContext.CreateService<IAgentServer>();
+            await _agentServer.ConnectAsync(connection);
+        }
+    }
+
+    public sealed class BuildReleasesAgentConfigProvider : ConfigurationProvider, IConfigurationProvider
+    {
+        public string ConfigurationProviderType
+            => Constants.Agent.AgentConfigurationProvider.BuildReleasesAgentConfiguration;
+
+        public void InitConnection(IAgentServer agentServer)
+        {
+            InitializeServerConnection(agentServer);
+        }
+
+        public void UpdateAgentSetting(AgentSettings settings)
+        {
+            // No implementation required
+        }
+
+        public string GetServerUrl(CommandSettings command)
+        {
+            return command.GetUrl(false);
+        }
+
+        public async Task<int> GetPoolId(CommandSettings command)
+        {
+            int poolId = 0;
+            string poolName;
+            while (true)
+            {
+                poolName = command.GetPool();
+                try
+                {
+                    poolId = await GetPoolIdAsync(poolName);
+                }
+                catch (Exception e) when (!command.Unattended)
+                {
+                    _term.WriteError(e);
+                }
+
+                if (poolId > 0)
+                {
+                    break;
+                }
+
+                _term.WriteError(StringUtil.Loc("FailedToFindPool"));
+            }
+            return poolId;            
+        }
+
+        public Task<TaskAgent> UpdateAgentAsync(int poolId, TaskAgent agent)
```

Comment:
you don't have to call the method in base class, just make the method in base as public virtual, then Machinegrouppriovider just over write the implementation.